addons/account/models/account_journal_dashboard.py

```diff
@@ -437,17 +439,45 @@ def _fill_sale_purchase_dashboard_data(self, dashboard_data):
             "account_move.invoice_date",
             "account_move.company_id",
         ]
-        query, params = sale_purchase_journals._get_open_bills_to_pay_query().select(*field_list)
+        payment_field_list = [
+            "account_move_line.journal_id",
+            "account_move_line.move_id",
+            "account_move_line.balance AS amount_total_company",
+        ]
+        # DRAFTS
+        query, params = sale_purchase_journals._get_draft_bills_query().select(*bills_field_list)
         self.env.cr.execute(query, params)
-        query_results_to_pay = group_by_journal(self.env.cr.dictfetchall())
+        query_results_drafts = group_by_journal(self.env.cr.dictfetchall())
 
-        query, params = sale_purchase_journals._get_draft_bills_query().select(*field_list)
+        # WAITING BILLS AND PAYMENTS
+        query, params = sale_purchase_journals._get_open_bills_to_pay_query().select(*bills_field_list)
         self.env.cr.execute(query, params)
-        query_results_drafts = group_by_journal(self.env.cr.dictfetchall())
+        query_results_bills_to_pay = group_by_journal(self.env.cr.dictfetchall())
+        if purchase_journals:
+            query, params = purchase_journals._get_open_payments_query().select(*payment_field_list)
+            self.env.cr.execute(query, params)
+            query_results_payments_to_pay = group_by_journal(self.env.cr.dictfetchall())
+        query_results_to_pay = {}
+        for journal in sale_purchase_journals:
+            if journal.type == "purchase":
+                query_results_to_pay[journal.id] = query_results_payments_to_pay[journal.id] or query_results_bills_to_pay[journal.id]
```

Comment:
Well, it does. For some journals, it seems that the aml query doesn't get the right totals while the move one does. I probably don't know enough about specific use cases to understand why but a test has been written for such a case and without that backup, the sum_waiting and sum_late end up being wrong.

Reply:
> it seems that the aml query doesn't get the right totals while the move one does

What if the query returns an amount that is not right but that is not zero then?

It's the concept of the `or` that makes me think it has to be wrong. It is right or it is not. Maybe you might need to sum both or something like that?

Anyways, needs a test.

Reply:
If the query retrieved a value, it is right. If it didn't, the other query got it right. Depending on the journal, the aml query will retrieve something or nothing at all. The final result is not some pieces of information coming from one query and some from the other. It is either one or the other, there is no "merge" of data with that `or`. 
I could replace:
```py
for journal in sale_purchase_journals:
            if journal.type == "purchase":
                query_results_to_pay[journal.id] = query_results_payments_to_pay[journal.id] or query_results_bills_to_pay[journal.id]
            else:
                query_results_to_pay[journal.id] = query_results_bills_to_pay[journal.id]
```
by
```py
for journal in sale_purchase_journals:
            if journal.type == "purchase" and query_results_payments_to_pay:
                query_results_to_pay[journal.id] = query_results_payments_to_pay[journal.id]
            else:
                query_results_to_pay[journal.id] = query_results_bills_to_pay[journal.id]
```
if it makes more sense to you.

For the testing part, there is already a test checking the values on the dashboard in odoo/addons/account/tests/test_account_journal_dashboard.py. I didn't modify it because the behavior did not change, We still check the total amount of the bills that are waiting for payment/overdue and those amount don't change based on whether you look at the moves or the amls. The only part of the behavior that is modified is the following : Previously, we were counting the number of moves that were involved in the total and now we count the number of amls that are involved. I can add a test for that part but it seems to me that if the total amount is right, then it means that every line that needed to be taken into account has been so testing the total is sufficient. Now that you know, I let it be your call.